Speed up OperatorEntry construction by avoiding updateDispatchTableFull_

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151850
* #151849
* #151810
* #151807
* #151806
* #151805
* #151804
* #151803
* #151802
* #151801
* #151800
* __->__ #151682

The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

swolchok committed:
Speed up OperatorEntry construction by avoiding updateDispatchTableFull_

The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

[ghstack-poisoned]

swolchok committed:
Update on "Speed up OperatorEntry construction by avoiding updateDispatchTableFull_"


The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -569,7 +687,7 @@ std::string OperatorEntry::dumpComputedTable() const {
   // Need to handle Undefined separately, because its a runtime key that can't be represented
   // in a DispatchKeySet.
   std::vector<DispatchKey> runtime_keys = {DispatchKey::Undefined};
```
Can reserve the std::vector size?

swolchok replied:
this looks like a debugging function, probably not worth the attention beyond the blanket find/replace I did. if I were to pay attention to it, I would just extract the loop body into a lambda and then call it first for undefined before looping over allDispatchKeysInFullset

swolchok committed:
Update on "Speed up OperatorEntry construction by avoiding updateDispatchTableFull_"


The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

[ghstack-poisoned]

swolchok committed:
Update on "Speed up OperatorEntry construction by avoiding updateDispatchTableFull_"


The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

[ghstack-poisoned]

swolchok commented:
holding this one until Tuesday per request from @bdhirsh

swolchok committed:
Update on "Speed up OperatorEntry construction by avoiding updateDispatchTableFull_"


The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

[ghstack-poisoned]

malfet commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -17,6 +22,48 @@ namespace {
 }
 
+static const std::vector<DispatchKey>& allDispatchKeysInFullSet() {
+  static const auto result = ([]() {
+    std::vector<DispatchKey> vec;
+    for (const auto dispatch_key: DispatchKeySet(DispatchKeySet::FULL)) {
+      vec.push_back(dispatch_key);
+    }
+    return vec;
+  })();
```
Nit (or point me to some doc that saying that extra parenthesis are good for lambda definitions 
```suggestion
  static const auto result = []() {
    std::vector<DispatchKey> vec;
    for (const auto dispatch_key: DispatchKeySet(DispatchKeySet::FULL)) {
      vec.push_back(dispatch_key);
    }
    return vec;
  }();
```

malfet commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -17,6 +22,48 @@ namespace {
+// entry is the DispatchKey that the corresponding index in the
+// dispatch table represents.
+static const auto& getDispatchTableIndexToKey() {
+  static const auto result = ([]() {
```
Why?
```suggestion
  static const auto result = []() {
```

malfet commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -17,6 +22,48 @@ namespace {
+      TORCH_INTERNAL_ASSERT(arr[index] != DispatchKey::Undefined, "missing dispatch key at index ", index);
+    }
+    return arr;
+  })();
```
Nit
```suggestion
  }();
```

malfet commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -17,6 +22,48 @@ namespace {
+    using result_type = std::array<DispatchKey, c10::num_runtime_entries>;
+    result_type arr;
+    arr.fill(DispatchKey::Undefined);
+    const auto update_array_entry = [](result_type& arr_, DispatchKey dk) {
+      const auto index = getDispatchTableIndexForDispatchKey(dk);
+      TORCH_INTERNAL_ASSERT(arr_.at(index) == DispatchKey::Undefined);
+      arr_.at(index) = dk;
+    };
+    for (const auto dk_outer: allDispatchKeysInFullSet()) {
+      update_array_entry(arr, dk_outer);
```
Why do you need a lambda there?
```suggestion
    for (const auto dk_outer: allDispatchKeysInFullSet()) {
      const auto index = getDispatchTableIndexForDispatchKey(dk);
      TORCH_INTERNAL_ASSERT(arr.at(index) == DispatchKey::Undefined);
      arr.at(index) = dk;
```

swolchok replied:
whoops, used to be called multiple times

swolchok committed:
Update on "Speed up OperatorEntry construction by avoiding updateDispatchTableFull_"


The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

[ghstack-poisoned]

swolchok committed:
rebase and minor fixes as requested on "Speed up OperatorEntry construction by avoiding updateDispatchTableFull_"


The purpose of the updateDispatchTableFull_ call is, according to the comment, just to pick up fallback kernels if there are any. We can implement that directly more efficiently.

Differential Revision: [D73129447](https://our.internmc.facebook.com/intern/diff/D73129447/)

[ghstack-poisoned]

malfet commented:
@pytorchbot merge
